enhancements/network/primary-user-defined-networks-for-virtualization-workloads.md

```diff
@@ -0,0 +1,568 @@
+the established TCP connections to be severed when the VM is migrated.
+
+This requires an enhancement on passt.
+
+TODO: link the ticket to get an enhancement
+
+This is the long-term plan for extending networking from the pod interface to
+the VM; assuming preserving the established TCP connections is an absolute must
+we could instead go for the [bridge binding alternative](#binding-mechanism) in
+4.18, and migrate to a version of passt that preserves the TCP connections in a
+future release / Z-stream.
+
+### Persisting VM IP addresses during the migration
+
+OpenShift already features the ability of providing persistent IP addresses
+for layer2 **secondary** user defined networks. It relies on the
+[IPAMClaim CRD](https://github.com/k8snetworkplumbingwg/ipamclaims/blob/48c5a915da3b67f464a4e52fa50dbb3ef3547dcd/pkg/crd/ipamclaims/v1alpha1/types.go#L23)
+to "tie" the allocation of the IP address to the VM.
+In short, OpenShift Virtualization creates an `IPAMClaim` resource for each
+interface in the VM the user wants to be persistent, and then instructs
+the CNI plugin of which claim to use to persist the IP for the VM using an
+attribute defined in the k8snetworkplumbingwg network-selection-elements (the
+`k8s.v1.cni.cncf.io/networks` annotation). This protocol is described in depth
+in
+[the persistent IPs for virt workloads enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/network/persistent-ips-sdn-secondary-networks.md#creating-a-virtual-machine).
+
+We want to re-use this mechanism to implement persistent IPs for UDN networks;
+meaning OpenShift Virtualization will create the `IPAMClaim` and will instruct
+OVN-Kubernetes of which claim to use. Since for primary UDNs we do **not**
+define the network-selection-elements, we need to use a new annotation to pass
+this information along.
+
+The proposed annotation value is `k8s.ovn.org/ovn-udn-ipamclaim-reference`.
+
+These persistent IPs will be cleaned up by the Kubernetes garbage collector
+once the VM to which they belong is removed.
+
+All the other
+[work required in OVN-Kubernetes](https://github.com/trozet/enhancements/blob/941f5c6391830d5e4a94e65d742acbcaf9b8eda9/enhancements/network/user-defined-network-segmentation.md#pod-egress)
+to support live-migration was already implemented as part of the epic
+implementing UDN.
+
+### Allow user to configure the VMs interface desired IP address
+
+We will also rely on the `IPAMClaim` CRD to provide the VM owner a way to
+specify the IP addresses for the UDN interface of their VMs. This way, we
+do not require to update the KubeVirt API, which is something we have faced
+extreme resistance in the past.
+
+Keep in mind that up to now the IPAMClaims would be created by KubeVirt
+directly; we need to allow the VM user to create them, specifying in its `spec`
+the desired IP addresses for the VM, and when creating the VM, point it to the
+respective `IPAMClaim` using an annotation (on the VM).
```

Comment:
Aint there will be an annotation on the VM as we spoke offline iirc ?
Requesting the user to create the claim, makes it prone for mistakes,
he will need to have the right owner Ref UID and so on, it is not comfortable user experience
he might also forget it

Reply:
The owner will be st by ipam controller, as it's done today.

Reply:
This is not what the doc says
"we need to allow the VM user to create them, specifying in its `spec`
the desired IP addresses for the VM, and when creating the VM, point it to the
respective `IPAMClaim` using an annotation (on the VM)."
and i think the problem is deeper than that (i need to think offline in order to provide examples)

more straight forward implementation will be that the VM / VMI has annotation
'xyz/persistentip: "1.2.3.4"
and the rest is almost agnostic beside reading this and adding it to the spec
this way user doesnt need to do anything about claims, cant have mistakes.

Reply:
I would suggest having user experience focus design, while keeping the code changes as low as possible
This way user need to know less, and maintain less CRDS

User adds `xyz/persistentip: "1.2.3.4"`  to the VM.VMI template
VMI Controller upon CREATE event, read the VMI anyhow today, it see this annotation on the VMI
it creates in the same flow as today the claim, and just add the requested IP to the IPAMClaim Spec
(this what Miguel presented offline iirc, and best imo)
Nothing beside need to be done wrt kubevirt / ipam-ext
(just OVN side need to play with it, as the doc says on other section, but this is trasparent)

Reply:
Sure, just adding alternatives here, whatever fits better.

Reply:
I would urge to keep the IPAMClaim API as close to [DRA](https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/) API as possible, so once DRA graduates, we can allow transition to it. With DRA, it is the Pod pointing at the Claim.